src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/FuseFunctionsPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Functions/FunctionFactory.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class FuseFunctionsMatcher : public InDepthQueryTreeVisitor<FuseFunctionsMatcher>
+{
+public:
+
+    static bool matchFunctionName(const String & name)
+    {
+        return name == "count" || name == "sum" || name == "avg";
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !function_node->isAggregateFunction() || !matchFunctionName(function_node->getFunctionName()))
+            return;
+
+        auto argument_hash = function_node->getArgumentsNode()->getTreeHash();
```

Comment:
I would prefer to check that we have single argument, and compute hash only for that single argument.